uobtheatre/bookings/models.py

```diff
@@ -74,6 +79,66 @@ class Meta:
         ]
 
 
+class BookingQuerySet(QuerySet):
+    """QuerySet for bookings"""
+
+    def annotate_checked_in(self):
+        return self.annotate(checked_in=BoolAnd("tickets__checked_in"))
+
+    def annotate_not_checked_in(self):
+        return self.annotate(count=models.Count("tickets")).annotate(
+            checked_in_count=models.Count(
+                Case(When(tickets__checked_in=True, then=Value(1)))
+            )
+        )
+
+    def annotate_checked_in_proportion(self):
+        return self.annotate_not_checked_in().annotate(
+            proportion=Case(
+                When(Q(count=0), then=Cast(0, FloatField())),
+                default=Cast(F("checked_in_count"), FloatField())
+                / Cast(F("count"), FloatField()),
+            )
+        )
+
+    def checked_in(self, bool_val=True):
```

Comment:
Can you type function definitions aswell plz?

Reply:
yup